docs(ecosystem-wg): May 1 website incident report (#280)
electron/governance: wg-ecosystem/incidents/2020-05-01.md

# Post-incident report for 1 May 2020

> ## What's happened

Reviewer: ```suggestion
## What happened
```

> 23 April after the default module updating, the website is unable to stable deploy. The fast reaction team disables the automatic deploys and reverts the website to the last good commit.

Reviewer: ```suggestion
On April 23, after some dependencies were updated, the website was unable to be deployed successfully. The Ecosystem WG disabled the automatic deploys and reverted the website to the last good commit.
```

## First investigation (around 0:00 UTC May 1)

> 1 May we started the investigation of that errors. We created 2 deploys, the first is `master` branch build with the latest changes, the second deploy is with a potential fix but that also failed. This investigation doesn't give more information since we are not able to find the root cause of errors. The big problem is that the crash only reproducible on `master` deploy, what's creates big problems with debugging.

Reviewer: ```suggestion
We started the investigation of the errors. We created 2 deploys: a `master` branch build with the latest changes, and a build with a potential fix - both failed. These tests did not result in any new information. The main problem was that the crash was only reproducible on the `master` deploy, which made debugging difficult.
```

## Further Investigation (Morning of May 1 for Europe)

> The future investigation continued by starting the isolated Heroku instance of a website. It's been created to read build logs, and easy debugging of error. The build log gives more information about what's Node version and dependencies we take under the building.

Reviewer: ```suggestion
The investigation continued by starting an isolated Heroku instance of the website. It was created to read build logs, and to more easily debug errors. The build log gave more information about the Node version and the dependencies used.
```

> The build logs give info that we use Node 14 when deploying from `master`, what we not test, and not plan to use for production deploys currently. The problem here it's that node version range says that we can use Node >= 12.

Reviewer: ```suggestion
The build logs indicated that we used Node 14 when deploying from `master`, which is different from the CI environment, and not what we planned to use for production deploys at this time. The issue is that `package.json` specifies the Node version range to `>= 12`.
```

> We don't took that version in the first since thinking the Heroku picks the Node 12 as they LTS but that not true. The errors that are we have seen 23 April it's that the deploy Heroku takes the Node 14 when it's not ready and tested on our CI processes.

Reviewer: ```suggestion
We didn't initially investigate the Node version, as we thought that Heroku picks Node 12 by default, as it is the LTS version (which is not the case).
```

> At around 19:45 UTC, we merge the hotfix patch what tested in the isolated instance and don't give any errors or crashes. This commit is successfully deployed, and we don't see any degradation after them.

Reviewer: ```suggestion
At around 19:45 UTC, we merged the hotfix patch that was tested in the isolated instance. This commit was successfully deployed, and there has not been any further downtime related to this problem since.
```

> ## Root cause of errors with deploying with Node 14

Reviewer: ```suggestion
## Root cause: deploying with Node 14
```

> We are unable to affirm these can be the root cause since unable to fully reproduce it, also the logs of the crash are empty and that creates problems of choosing the right version. These version is most probably:

Reviewer: ```suggestion
We are unable to confirm that this is the root cause since we were unable to fully reproduce it, and the logs of the crash are empty. The most likely reasons why Node 14 could not deploy the website are:
```

> We are unable to affirm these can be the root cause since unable to fully reproduce it, also the logs of the crash are empty and that creates problems of choosing the right version. These version is most probably:

Reviewer 2: Leaving a note here to update the cause with https://github.com/nodejs/node/issues/33283

> - Building `node-sass` instead of using the prebuilt binaries.

Reviewer: ```suggestion
- Building `node-sass` from source instead of using the prebuilt binaries.
```

## Further Investigation (Morning of May 1 for Europe)

The investigation continued by starting an isolated Heroku instance of the website. It was created to read build logs, and to more easily debug errors. The build log gave more information about the Node version and the dependencies used.

The build logs indicated that we used Node 14 when deploying from `master`, which is different from the CI environment, and not what we planned to use for production deploys at this time. The issue is that `package.json` specifies the Node version range to `>= 12`.

We didn't initially investigate the Node version, as we thought that Heroku picks Node 12 by default, as it is the LTS version (which is not the case).

At around 19:45 UTC, we merged the hotfix patch that was tested in the isolated instance. This commit was successfully deployed, and there has not been any further downtime related to this problem since.

## Root cause: deploying with Node 14

> > We found reproduction and root cause of the crash and created issue in [Node.js repository](https://github.com/nodejs/node/issues/33283).
> >
> > We update the post when we have more information.
> > That section not be updated since at the moment of writing we didn't know the root cause.
> 
> We are unable to confirm that this is the root cause since we were unable to fully reproduce it, and the logs of the crash are empty. The most likely reasons why Node 14 could not deploy the website are:
> 
> - Building `node-sass` from source instead of using the prebuilt binaries.
> - Some module has errors when running Node 14.

Reviewer: ```suggestion
We found a way to reproduce the crash and created an issue in the [Node.js repository](https://github.com/nodejs/node/issues/33283).

We will update this report when we have more information.
```
